src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -24,25 +25,71 @@
 /* controlled via GUC, used mostly for testing */
 bool LogIntermediateResults = false;
 
+
+static List * FindSubPlansUsedInNode(Node *node, SubPlanAccessType accessType);
 static void AppendAllAccessedWorkerNodes(IntermediateResultsHashEntry *entry,
 										 DistributedPlan *distributedPlan,
 										 int workerNodeCount);
+static void AppendAllWorkerNodes(IntermediateResultsHashEntry *entry);
 static List * FindAllRemoteWorkerNodesUsingSubplan(IntermediateResultsHashEntry *entry);
 static List * RemoveLocalNodeFromWorkerList(List *workerNodeList);
 static void LogIntermediateResultMulticastSummary(IntermediateResultsHashEntry *entry,
 												  List *workerNodeList);
-static bool UsedSubPlansEqual(UsedDistributedSubPlan *left,
-							  UsedDistributedSubPlan *right);
-static UsedDistributedSubPlan * UsedSubPlanListMember(List *list,
-													  UsedDistributedSubPlan *usedPlan);
+
+
+/*
+ * FindSubPlanUsages finds the subplans used in the master query and the
+ * job query and returns them as a combined list of UsedDistributedSubPlan
+ * structs.
+ *
+ * The list may contain duplicates if the subplan is referenced multiple
+ * times.
+ */
+List *
+FindSubPlanUsages(DistributedPlan *plan)
+{
+	List *localSubPlans = NIL;
+	List *remoteSubPlans = NIL;
+
+	if (plan->masterQuery != NULL)
+	{
+		localSubPlans = FindSubPlansUsedInNode((Node *) plan->masterQuery,
+											   SUBPLAN_ACCESS_LOCAL);
+	}
+
+	if (plan->workerJob != NULL)
+	{
+		/*
+		 * Mark the subplans as needed on remote side. Note that this decision is
+		 * revisited on execution, when the query only consists of intermediate
+		 * results.
+		 */
+		remoteSubPlans = FindSubPlansUsedInNode((Node *) plan->workerJob->jobQuery,
+												SUBPLAN_ACCESS_REMOTE);
+	}
+
+	if (plan->insertSelectQuery != NULL)
```

Comment:
`Assert(plan->workerJob == NULL)`

Reply:
Is there a reason that it is not `else if (plan->insertSelectQuery != NULL)`?  Now it feels a little unusual to update `remoteSubPlans ` in two different places.